extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java

```diff
@@ -0,0 +1,241 @@
+     * @param queryExecutor - The {@link QueryExecutor} that starts/creates
+     *        queries when a CREATED query is found in the
+     *        {@link QueryChangeLog}. (not null)
+     * @param sources - The {@link QueryChangeLogSource}s the sources of
+     *        QueryChangeLogs. (not null)
+     * @param queryRepos - The set of {@link QueryRepository}s that will notify
+     *        this manager when {@link ChangeLogEntry}s occur. (not null)
+     * @param scheduler - The {@link Scheduler} to use throughout
+     */
+    public QueryManager(final String ryaInstanceName, final QueryExecutor queryExecutor,
+            final Map<String, QueryChangeLogSource> sources, final Set<QueryRepository> queryRepos,
+            final Scheduler scheduler) {
+        this.ryaInstanceName = requireNonNull(ryaInstanceName);
+        this.sources = requireNonNull(sources);
+        this.queryExecutor = requireNonNull(queryExecutor);
+        this.queryRepos = requireNonNull(queryRepos);
+        this.scheduler = requireNonNull(scheduler);
+
+        // subscribe to the repos and sources to be notified of changes.
+        queryRepos.forEach(repo -> repo.subscribe(new QueryManagerQueryChange()));
+        sources.forEach((ryaInstance, source) -> source.subscribe(new QueryManagerSourceListener()));
+
+        // create query cache.
+        queriesCache = new HashMap<>();
+    }
+
+    /**
+     * Starts running a query.
+     *
+     * @param query - The query to run.(not null)
+     */
+    private void runQuery(final StreamsQuery query) {
+        requireNonNull(query);
+        LOG.trace("Starting Query: " + query.getSparql());
+        if (!queryExecutor.isRunning()) {
+            LOG.debug("Query Executor wasn't running, starting now.");
+            queryExecutor.startAndWait();
+        }
+
+        try {
+            queryExecutor.startQuery(ryaInstanceName, query);
+        } catch (final QueryExecutorException e) {
+            LOG.error("Failed to start query.", e);
+        }
+    }
+
+    /**
+     * Stops the specified query from running.
+     *
+     * @param queryId - The ID of the query to stop running. (not null)
+     */
+    private void stopQuery(final UUID queryId) {
+        requireNonNull(queryId);
+
+        LOG.trace("Stopping query: " + queryId.toString());
+        if (!queryExecutor.isRunning()) {
+            throw new IllegalStateException("Query Executor must be started before queries can be managed.");
+        }
+
+        try {
+            queryExecutor.stopQuery(queryId);
+        } catch (final QueryExecutorException e) {
+            LOG.error("Failed to stop query.", e);
+        }
+    }
+
+    @Override
+    protected void startUp() throws Exception {
+        LOG.trace("Starting Query Manager.");
+        queryRepos.forEach(repo -> repo.startAndWait());
+        sources.forEach((ryaInstance, source) -> source.startAndWait());
+    }
+
+    @Override
+    protected void shutDown() throws Exception {
```

Comment:
Might want to use stop and wait calls instead of just stop to prevent this method from returning until everything the class depends on finishes stopping.

Reply:
done